sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/analysis/ResolveDefaultColumns.scala

```diff
@@ -0,0 +1,354 @@
+/*
+ * Licensed to the Apache Software Foundation (ASF) under one or more
+ * contributor license agreements.  See the NOTICE file distributed with
+ * this work for additional information regarding copyright ownership.
+ * The ASF licenses this file to You under the Apache License, Version 2.0
+ * (the "License"); you may not use this file except in compliance with
+ * the License.  You may obtain a copy of the License at
+ *
+ *    http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package org.apache.spark.sql.catalyst.analysis
+
+import org.apache.spark.sql.AnalysisException
+import org.apache.spark.sql.catalyst.TableIdentifier
+import org.apache.spark.sql.catalyst.analysis.ResolveDefaultColumns._
+import org.apache.spark.sql.catalyst.catalog._
+import org.apache.spark.sql.catalyst.expressions._
+import org.apache.spark.sql.catalyst.optimizer.ConstantFolding
+import org.apache.spark.sql.catalyst.parser.{CatalystSqlParser, ParseException}
+import org.apache.spark.sql.catalyst.plans.logical._
+import org.apache.spark.sql.catalyst.rules.Rule
+import org.apache.spark.sql.catalyst.trees.AlwaysProcess
+import org.apache.spark.sql.internal.SQLConf
+import org.apache.spark.sql.types._
+
+/**
+ * This is a rule to process DEFAULT columns in statements such as CREATE/REPLACE TABLE.
+ *
+ * Background: CREATE TABLE and ALTER TABLE invocations support setting column default values for
+ * later operations. Following INSERT, and INSERT MERGE commands may then reference the value
+ * using the DEFAULT keyword as needed.
+ *
+ * Example:
+ * CREATE TABLE T(a INT DEFAULT 4, b INT NOT NULL DEFAULT 5);
+ * INSERT INTO T VALUES (1, 2);
+ * INSERT INTO T VALUES (1, DEFAULT);
+ * INSERT INTO T VALUES (DEFAULT, 6);
+ * SELECT * FROM T;
+ * (1, 2)
+ * (1, 5)
+ * (4, 6)
+ *
+ * @param catalog the catalog to use for looking up the schema of INSERT INTO table objects.
+ * @param insert the enclosing INSERT statement for which this rule is processing the query, if any.
+ */
+case class ResolveDefaultColumns(
+    catalog: SessionCatalog, insert: Option[InsertIntoStatement] = None) extends Rule[LogicalPlan] {
+  override def apply(plan: LogicalPlan): LogicalPlan = plan.resolveOperatorsWithPruning(
+    AlwaysProcess.fn, ruleId) {
+
+    case i@InsertIntoStatement(_, _, _, _, _, _)
+      if SQLConf.get.enableDefaultColumns &&
+        i.query.collectFirst { case u: UnresolvedInlineTable => u }.isDefined =>
+      val newQuery = ResolveDefaultColumns(catalog, Some(i)).apply(i.query)
```

Comment:
Note that `plan.resolveOperatorsWithPruning` will transform all the operators in `i` including query recursively.
Shall we move the methods `addMissingDefaultColumnValues` and `replaceExplicitDefaultColumnValues` to `object ResolveDefaultColumns` so that we don't need to create class `ResolveDefaultColumns` recursively.
The same applies to the following code.

Reply:
Good question: this is actually the intent to transform recursively, specifically only on the subplan in `i.query`. In this way, the following case matching against `UnresolvedInlineTable` may trigger anywhere that the corresponding `values` list appears in that subplan (e.g. `insert into t select * from values(...)` or more complex cases ending in `values`). I left a comment explaining this here. Otherwise in the more complex cases, the analyzer simply errors that `default` was not found in those cases.

If this SG to you, we can leave it as-is. Otherwise, if you prefer to only match the `i.query` against (`UnresolvedInlineTable` or `Project <- UnresolvedInlineTable`), let me know and I can make that change. We could possibly add another `case` then to `find` if any unresolved `default` references exist in any nested `values`, and return a more specific error message in that case to help guide the user.